otherlibs/systhreads/st_posix.h

```diff
@@ -197,12 +197,26 @@ typedef pthread_mutex_t * st_mutex;
 static int st_mutex_create(st_mutex * res)
 {
   int rc;
-  st_mutex m = caml_stat_alloc_noexc(sizeof(pthread_mutex_t));
-  if (m == NULL) return ENOMEM;
-  rc = pthread_mutex_init(m, NULL);
-  if (rc != 0) { caml_stat_free(m); return rc; }
+  pthread_mutexattr_t attr;
+  st_mutex m;
+
+  rc = pthread_mutexattr_init(&attr);
+  if (rc != 0) goto error1;
+  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
+  if (rc != 0) goto error2;
+  m = caml_stat_alloc_noexc(sizeof(pthread_mutex_t));
+  if (m == NULL) { rc = ENOMEM; goto error2; }
+  rc = pthread_mutex_init(m, &attr);
+  if (rc != 0) goto error3;
+  pthread_mutexattr_destroy(&attr);
   *res = m;
   return 0;
+error3:
+  caml_stat_free(m);
+error2:
+  pthread_mutexattr_destroy(&attr);
+error1:
+  return rc;
```

Comment:
Possible version with fewer labels...
```suggestion
  if (rc != 0) return rc;
  rc = pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_ERRORCHECK);
  if (rc != 0) goto finish;
  m = caml_stat_alloc_noexc(sizeof(pthread_mutex_t));
  if (m == NULL) { rc = ENOMEM; goto finish; }
  rc = pthread_mutex_init(m, &attr);
  if (rc == 0) { *res = m; goto finish; }
  caml_stat_free(m);
finish:
  pthread_mutexattr_destroy(&attr);
  return rc;
```

Reply:
I'll read this again, but I sort of like the "state machine" with the 3 error labels.

Reply:
Eliminating the double `pthread_mutexattr_destroy` seemed nicer (which can be done without removing all the labels!)